### PR TITLE
Laravel 8.x Shift

### PR DESCRIPTION
**Pull request by driesvints**
This pull request includes the changes for upgrading to Laravel 8.x. Feel free to commit any additional changes to the `shift-35839` branch.

**Before merging**, you need to:

- Checkout the `shift-35839` branch
- Review **all** pull request comments for additional changes
- Update your dependencies for Laravel 8
- Run `composer update` (if the scripts fail, add `--no-scripts`)
- Thoroughly test your application ([no tests?](https://confidentlaravel.com))

If you need help with your upgrade, check out the [Human Shifts](https://laravelshift.com/human-shifts). You may also join the [Shifty Coders](https://laravelshift.com/shifty-coders) Slack workspace to level-up your Laravel skills.

### CODE REVIEW

**Comment by driesvints** [comment]
:x: Shift could not upgrade the following files since they differed from the default Laravel version. You will need to compare these files against the [default Laravel 8 versions](https://github.com/laravel-shift/laravel-8.x/) and merge any changes:

- [ ] public/index.php

**Comment by driesvints** [comment]
:information_source: Laravel 8 streamlined the _Exception Handler_ into a single `register` method to make it easier to add custom logic around rendering or reporting errors within your application.

While this change is not required, Shift recommends comparing yours against the [Laravel 8.x version](https://github.com/laravel-shift/laravel-8.x/blob/master/app/Exceptions/Handler.php) as an opportunity to reduce code and be fully up-to-date.

**Comment by driesvints** [comment]
:x: To match the other _auth_ middleware, Laravel 8 now allows multiple guards for the `RedirectIfAuthenticated` middleware.

Shift attempted to automate these changes, but detected customizations. You will need to compare your `app/Http/Middleware/RedirectIfAuthenticated.php`  with the [default Laravel 8 version](https://github.com/laravel-shift/laravel-8.x/blob/master/app/Http/Middleware/RedirectIfAuthenticated.php) to finalize the upgrade.

**Comment by driesvints** [comment]
:information_source: The `RouteServiceProvider` no longer prefixes the controller namespace by default in Laravel 8. While you may still set this property, you are encouraged to register your routes using static `::class` references.

You may automate the conversion to static class references using the [Shift Workbench](https://laravelshift.com/workbench).

**Comment by driesvints** [comment]
:x: Laravel 8 condenses the `RouteServiceProvider` as well as configured the [new rate limiters](https://laravel.com/docs/8.x/routing#rate-limiting). Shift was unable to automate this change. You should compare yours against the [Laravel 8.x version](https://github.com/laravel-shift/laravel-8.x/blob/master/app/Providers/RouteServiceProvider.php) and merge any changes.

**Comment by driesvints** [comment]
:information_source: Laravel 8 introduced class based factories. Shift [automated this change](https://github.com/laravelio/laravel.io/pull/563/commits/d9eddadd25573bbc93a9970ca9144dcdd4bbd7f0?w=1). However, if you were using a community package for factories or other custom code, you should review your factories for additional changes.

If you encounter an issue, you may revert this commit and temporarily install the [laravel/legacy-factories](https://github.com/laravel/legacy-factories) package. You may then automate the conversion to class based factories using the [Shift Workbench](https://laravelshift.com/workbench).

**Comment by driesvints** [comment]
:information_source: Laravel 8 added a `uuid` column to the `failed_jobs` table to support the new [job batching](https://laravel.com/docs/8.x/queues#job-batching) features.

If your application is currently using the `failed_jobs` table, you may generate a custom migration for this command by running:

```sh
php artisan make:migration AddUuidToFailedJobsTable
```

Then copy and paste the contents from [Shift's `AddUuidToFailedJobsTable` class](https://github.com/laravel-shift/laravel-8.x/blob/master/database/migrations/2020_09_08_000000_add_uuid_to_failed_jobs_table.php).

**Comment by driesvints** [comment]
:information_source: Laravel 8 added a new ENV variable of `LOG_LEVEL`. While the log level is defaulted in the configuration, you may take this opportunity to set it accordingly in your environments.

**Comment by driesvints** [comment]
:warning: Shift upgraded your configuration files by defaulting them and merging your _true customizations_. These include values which are not changeable through core `ENV` variables. This should make [maintaining your config files easier](https://jasonmccreary.me/articles/maintaining-laravel-config-files/).

You should [review this commit](https://github.com/laravelio/laravel.io/pull/563/commits/281103a8a81dc739ee5903086a97deeb5be29a72?w=1) for any additional customizations. If you have a lot of customizations, you may wish to undo this commit with `git revert` and make these [config file changes](https://github.com/laravel-shift/laravel-7.x/pull/1/commits/aff49348815635b46f7aa2e5c1d93504f3920bf6) manually.

**Comment by driesvints** [comment]
:information_source: Shift detected a dependency for `livewire/livewire`. Livewire 2 is now available. However, this requires [additional changes](https://laravel-livewire.com/docs/2.x/upgrading) to your application.

Shift did not update your dependency as it is compatible with Laravel 8.

**Comment by driesvints** [comment]
:information_source: Shift [updated your dependencies](https://github.com/laravelio/laravel.io/pull/563/commits/e2cb64644faf1f673e0a18776be9e03547d88377?w=1) for Laravel 8. While many of the popular packages are reviewed, you may have to update additional packages in order for your application to be compatible with Laravel 8.

Watch [dealing with dependencies](https://laravelshift.com/videos/update-incompatible-composer-dependencies) for tips on handling package incompatibilities.

**Comment by driesvints** [comment]
:warning: Laravel 8 now requires PHPUnit 9 which made some configuration changes to its config file. You may automate these changes by running `vendor/bin/phpunit --migrate-configuration` or comparing your `phpunit.xml` to the [Laravel 8 version](https://github.com/laravel-shift/laravel-8.x/blob/master/phpunit.xml) and merging any changes.

**Comment by driesvints** [comment]
:warning: The paginator now uses [Tailwind](https://tailwindcss.com/) for its default styling in Laravel 8. If you wish to continue using Bootstrap, you should add the following snippet to the `boot` method of your `AppServiceProvider`:

```php
\Illuminate\Pagination\Paginator::useBootstrap();
```

**Comment by driesvints** [comment]
:warning: Shift detected you are using a Laravel package like Horizon or Nova which may need to have its published assets regenerated after upgrading. Be sure to use `artisan` to _republish_ these assets as well as `php artisan view:clear` to avoid any errors.

**Comment by driesvints** [comment]
:tada: Congratulations, you're now running the latest version of Laravel.

The following _Shifts_ may be used ensure your application is fully upgraded and easier to maintain in the future:

- [Upgrade Checker](https://laravelshift.com/laravel-upgrade-checker) ensures your application is fully upgraded by detecting any outdated code.
- [Laravel Fixer](https://laravelshift.com/laravel-code-fixer) automates many of the suggestions found by the [_Laravel Linter_](https://laravelshift.com/opinionated-laravel-way-shift) and other _Shifts_.
- [Tests Generator](https://laravelshift.com/laravel-test-generator) intelligently generates model factories, HTTP Tests, and unit tests for your application to give you a head start on testing.

You may run these _Shifts_ and stay upgraded forever as well by subscribing to a [Shifty Plans](https://laravelshift.com/shifty-plans).

**Comment by joedixon** [comment]
@driesvints I think this is good to go

File: phpunit.xml [hunk 1]
```diff
@@ -1,34 +1,30 @@
...
+    <env name="APP_KEY" value="base64:NXoQgjw2ZlOxnGbo5ZRhYgTdM6xLYsgYElNAgcTQJkE="/>
+    <env name="CACHE_DRIVER" value="array"/>
+    <env name="DB_CONNECTION" value="sqlite"/>
+    <env name="DB_DATABASE" value=":memory:"/>
+    <env name="MAIL_MAILER" value="array"/>
+    <env name="SESSION_DRIVER" value="array"/>
+    <env name="QUEUE_CONNECTION" value="sync"/>
+    <env name="SCOUT_DRIVER" value="null"/>
+  </php>
 </phpunit>
```
Comment (driesvints) [inline phpunit.xml:30]: I'd like to revert the indentation here to make it the same as the base Laravel skeleton. Not sure what caused this.

File: app/Http/Middleware/RedirectIfAuthenticated.php [hunk 2]
```diff
@@ -10,15 +11,20 @@ class RedirectIfAuthenticated
     /**
      * Handle an incoming request.
      *
-     * @param  \Illuminate\Http\Request  $request
-     * @param  \Closure  $next
-     * @param  string|null  $guard
+     * @param \Illuminate\Http\Request $request
+     * @param Closure $next
+     * @param string[]|null ...$guards
+     *
```
Comment (driesvints) [inline app/Http/Middleware/RedirectIfAuthenticated.php:17]: Let's revert these docblock changes for now.

File: app/Providers/RouteServiceProvider.php [hunk 3]
```diff
@@ -2,71 +2,63 @@
...
-     * The path to the "home" route for your application.
+     * The controller namespace for the application.
      *
-     * @var string
+     * When present, controller route declarations will automatically be prefixed with this namespace.
+     *
+     * @var string|null
      */
-    public const HOME = '/dashboard';
+    // protected $namespace = 'App\\Http\\Controllers';
```
Comment (driesvints) [inline app/Providers/RouteServiceProvider.php:28]: You can totally remove this property.

File: composer.json [hunk 4]
```diff
@@ -5,53 +5,51 @@
...
         "spatie/laravel-robots-middleware": "^1.2",
         "thenpingme/laravel": "^2.2",
         "yarri/link-finder": "^2.5"
     },
     "require-dev": {
-        "barryvdh/laravel-debugbar": "^3.3",
-        "fzaninotto/faker": "^1.9",
-        "laravel/browser-kit-testing": "^6.0",
+        "barryvdh/laravel-debugbar": "^3.5",
+        "fzaninotto/faker": "^1.9.1",
```
Comment (driesvints) [inline composer.json:39]: Update this dependency to `fakerphp/faker`

File: database/migrations/2020_11_03_205735_add_uuid_to_failed_jobs_table.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('failed_jobs', function (Blueprint $table) {
+            $table->dropColumn('uuid');
+        });
+    }
```
Comment (driesvints) [inline database/migrations/2020_11_03_205735_add_uuid_to_failed_jobs_table.php:31]: Remove the down method.

File: database/migrations/2020_11_03_205735_add_uuid_to_failed_jobs_table.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class AddUuidToFailedJobsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
```
Comment (driesvints) [inline database/migrations/2020_11_03_205735_add_uuid_to_failed_jobs_table.php:13]: Remove the docblock here

**Comment by driesvints** [comment]
@joedixon just a few remarks.

File: phpunit.xml (hunk 1 above)
Comment (joedixon) [reply phpunit.xml:30]: On it. It was caused by the command to migrate to the new format required for PHPUnit 9

File: app/Http/Middleware/RedirectIfAuthenticated.php (hunk 2 above)
Comment (joedixon) [reply app/Http/Middleware/RedirectIfAuthenticated.php:17]: Do you also want me to revert the changes to the handle method?

**Comment by joedixon** [comment]
@driesvints pinging you here as I can't add request a review as it's your PR. Made all changes except the one I commented on above.

File: phpunit.xml [hunk 7]
```diff
@@ -1,25 +1,21 @@
...
-         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
-         bootstrap="vendor/autoload.php"
-         colors="true"
->
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" bootstrap="vendor/autoload.php" colors="true">
+    <coverage processUncoveredFiles="true">
+        <include>
+        <directory suffix=".php">./app</directory>
+        </include>
+    </coverage>
```
Comment (driesvints) [inline phpunit.xml:7]: Remove this part. It's unneeded.

File: phpunit.xml [hunk 8]
```diff
@@ -1,25 +1,21 @@
...
-            <directory suffix="Test.php">./tests/Unit</directory>
+        <directory suffix="Test.php">./tests/Unit</directory>
         </testsuite>
         <testsuite name="Integration">
-            <directory suffix="Test.php">./tests/Integration</directory>
+        <directory suffix="Test.php">./tests/Integration</directory>
         </testsuite>
         <testsuite name="Feature">
-            <directory suffix="Test.php">./tests/Feature</directory>
+        <directory suffix="Test.php">./tests/Feature</directory>
```
Comment (driesvints) [inline phpunit.xml:16]: Indentations are off here. I think it's best that we don't use a command to format this. Can we just revert the entirety here back to how the file was and only remove the filter block?

File: app/Http/Middleware/RedirectIfAuthenticated.php (hunk 2 above)
Comment (driesvints) [reply app/Http/Middleware/RedirectIfAuthenticated.php:17]: No it's fine.

Comment (joedixon) [reply app/Http/Middleware/RedirectIfAuthenticated.php:17]: I guess I'm being slow, but in that case, why revert the docblock? This matches the Laravel source

Comment (driesvints) [reply app/Http/Middleware/RedirectIfAuthenticated.php:17]: Indentation is off: https://github.com/laravel/laravel/blob/8.x/app/Http/Middleware/RedirectIfAuthenticated.php#L15-L17

**Comment by joedixon** [comment]
How's that?

**Comment by driesvints** [comment]
Perfect. Thanks Joe!

**Pull request merged**